### PR TITLE
Add Discv5 Scaffolding

### PR DESCRIPTION
**Pull request by carver**
- Rebase and squash commits for discv5 and TalkRequest
- Launch the portal p2p and json-rpc server side by side
- Updated CI to catch clippy and other compile warnings
- Resolve all the warnings, except dead code ones
- Ignoring the dead code warnings for now, plan to remove it later

Haven't yet been able to bond with a `ddht` peer. Will save that for another PR, so we don't have so much code waiting unmerged.

### CODE REVIEW

**Comment by carver** [comment]
@SamWilsn Trying to get this merged in so we're all working off the same base.

**Review by njgheorghita — APPROVED** [review]
Yeah, this looks nice! Though, tbh I learned more about rust reviewing this than I'd be able to constructively apply to the review itself...

File: src/alexandria/discovery.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+pub struct Discovery {
+    discv5: Discv5,
+    /// Indicates if the discv5 service has been started
+    pub started: bool,
+}
+
+impl Discovery {
+    pub fn new(config: Config) -> Result<Self, String> {
+        let enr_key = CombinedKey::generate_secp256k1();
```
[the remark this answers is not in the document]
Comment (carver) [reply src/alexandria/discovery.rs:38]: Agree, opened #15 to track.

File: src/alexandria/discovery.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        let enr = {
+            let mut builder = EnrBuilder::new("v4");
+            builder.ip(config.listen_address);
+            builder.udp(config.listen_port);
+            builder.build(&enr_key).unwrap()
+        };
+
+        info!("Starting discv5 service with enr: {:?}", enr);
+
+        info!("Self enr: {}", enr);
```
[the remark this answers is not in the document]
Comment (carver) [reply src/alexandria/discovery.rs:49]: Yeah, the `{:?}` and `{}` print out seriously different things (the encoded and decoded ENRs). But it's a fair point that this is best shown as a single line, with a little more clarity. Will do. :+1:

File: src/alexandria/discovery.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        let mut discv5 = Discv5::new(enr, enr_key, config.discv5_config)
+            .map_err(|e| format!("Failed to create discv5 instance: {}", e))?;
+
+        for enr in config.bootnode_enrs {
+            info!("Adding bootnode");
```
[the remark this answers is not in the document]
Comment (carver) [reply src/alexandria/discovery.rs:54]: Yeah. We can always drop it to `debug` later if it feels overly verbose.

File: src/alexandria/protocol.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    fn default() -> Self {
+        Self {
+            listen_port: 4242,
+            bootnode_enrs: vec![],
+            data_radius: U256::from(u64::MAX), //TODO better data_radius default?
+        }
+    }
+}
+
+pub const PROTOCOL: &str = "state-network";
```
[the remark this answers is not in the document]
Comment (carver) [reply src/alexandria/protocol.rs:32]: Ah, I think we should probably update the python one actually. "alexandria" should be the chain history one. Which is also a reminder to rename this module from alexandria to something like `statenet`.

**Pull request merged**